src/platform/Darwin/BleConnectionDelegateImpl.mm

```diff
@@ -280,21 +276,23 @@ - (BOOL)isConnecting
     return _currentMode == kConnecting;
 }
 
-- (void)resetTimer
+- (void)setupTimer:(uint64_t)timeout
 {
-    if ([self isConnecting]) {
-        auto timeout = static_cast<int64_t>(kScanningWithDiscriminatorTimeoutInSeconds * NSEC_PER_SEC);
-        dispatch_source_set_timer(_timer, dispatch_walltime(nullptr, timeout), DISPATCH_TIME_FOREVER, 5 * NSEC_PER_SEC);
-    } else if ([self isScanningWithoutDelegate]) {
-        auto timeout = static_cast<int64_t>(kScanningWithoutDelegateTimeoutInSeconds * NSEC_PER_SEC);
-        dispatch_source_set_timer(_timer, dispatch_walltime(nullptr, timeout), DISPATCH_TIME_FOREVER, 5 * NSEC_PER_SEC);
-    } else if ([self isScanning]) {
-        dispatch_source_cancel(_timer);
-    } else {
-        // It should not happens.
+    _timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, _workQueue);
```

Comment:
If we already had a time coming in here, do we need to cancel it somehow?  Or is the point that it should always be nil when we enter this function?

Reply:
I have added additional clean up code at the beginning of setupTimer